saltgui/static/scripts/routes/keys.js

```diff
@@ -6,23 +6,50 @@ class KeysRoute extends PageRoute {
     this.jobsLoaded = false;
 
     this._updateKeys = this._updateKeys.bind(this);
+    this._updateKeysFingerprint = this._updateKeysFingerprint.bind(this);
   }
 
   onShow() {
     const keys = this;
     return new Promise(function(resolve, reject) {
       keys.resolvePromise = resolve;
       if(keys.keysLoaded && keys.jobsLoaded) resolve();
-      keys.router.api.getMinions().then(keys._updateMinions);
       keys.router.api.getKeys().then(keys._updateKeys);
+      keys.router.api.getKeysFingerprint().then(keys._updateKeysFingerprint);
       keys.router.api.getJobs().then(keys._updateJobs);
       keys.router.api.getJobsActive().then(keys._runningJobs);
     });
   }
 
+  _updateKeysFingerprint(data) {
+    const keys = data.return[0].data.return;
+
+    for(const property of Object.keys(keys)) {
+      if(property === "local") continue;
+      const hosts = keys[property];
+      for(const hostname of Object.keys(hosts)) {
+        const fingerprint = hosts[hostname];
+        // why this is sometimes null?
+        const element = document.querySelector("#" + hostname);
```

Comment:
@erwindon very often I have race condition when `keys.router.api.getKeysFingerprint()` is retrieved quicker then `keys.router.api.getKeys()` and then I don't have yet pre-filled informations so `_updateKeysFingerprint` will not work. Do you know what I can do to force proper order (some kind of chaining maybe)?

Reply:
You can combine multiple functions using `Promise.all` and use a single handler to process both results at the same time (calling the 2 original handlers sequentially). See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all.
This should be generalized for all API functions on all pages so that none of the additional details are processed before the page structure has been set up.